File: v1/backend/storage/repos/9f9d7705-da3e-45fb-b4fd-3b342766baf8/tuck-in-tales-backend/src/models/memory.py
from uuid import UUID, uuid4
from datetime import date, datetime
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from enum import Enum


class AnalysisStatus(str, Enum):
    PENDING = "PENDING"
    ANALYZING = "ANALYZING"
    ANALYZED = "ANALYZED"
    CONFIRMED = "CONFIRMED"
    FAILED = "FAILED"


class MemoryCategory(str, Enum):
    CONTEXT = "context"
    NEW_CHARACTER = "new_character"
    IMPORTANT_EVENT = "important_event"
    FEELING_EMOTION = "feeling_emotion"
    WATCH_FOR = "watch_for"


class SuggestionType(str, Enum):
    NEW_CHARACTER = "new_character"
    UPDATE_CHARACTER_BIO = "update_character_bio"
    UPDATE_CHARACTER_VISUAL = "update_character_visual"
    UPDATE_CHARACTER_AVATAR = "update_character_avatar"


class MemorySuggestion(BaseModel):
    type: SuggestionType
    character_id: Optional[str] = None
    character_name: Optional[str] = None
    data: dict = Field(default_factory=dict)


class MemoryBase(BaseModel):
    text: Optional[str] = None
    date: date
    # family_id: UUID # Removed - Determined by auth
    # embedding: vector(1536) - Handled internally, not part of API model


class MemoryCreate(BaseModel):
    text: Optional[str] = None
    date: Optional[date] = None  # defaults to today in route


class MemoryUpdate(BaseModel):
    text: Optional[str] = None
    date: Optional[date] = None
    # family_id and embedding cannot be changed directly via API


class Memory(BaseModel):
    id: UUID
    family_id: UUID
    text: Optional[str] = None
    date: date
    photo_paths: List[str] = Field(default_factory=list)
    categories: List[str] = Field(default_factory=list)
    summary: Optional[str] = None
    linked_character_ids: List[str] = Field(default_factory=list)
    analysis_status: str = AnalysisStatus.PENDING
    raw_analysis: Optional[Any] = None
    suggestions: Optional[List[MemorySuggestion]] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        from_attributes = True


class MemoryConfirmRequest(BaseModel):
    categories: List[str] = Field(default_factory=list)
    summary: Optional[str] = None
    linked_character_ids: List[str] = Field(default_factory=list)


class ApplySuggestionRequest(BaseModel):
    suggestion_index: int
    approved: bool = True


class NewCharacterFromMemory(BaseModel):
    name: str
    bio: Optional[str] = None
    photo_index: Optional[int] = None  # which memory photo to copy as character photo
    face_x: Optional[float] = None  # 0-100, for cropping multi-person photos


class CreateCharactersFromMemoryRequest(BaseModel):
    characters: List[NewCharacterFromMemory]


# Models for RAG Search
class MemorySearchRequest(BaseModel):
    query: str
    match_threshold: float = 0.7
    match_count: int = 5


class MemorySearchResult(Memory):
    similarity: float

File: v1/backend/storage/repos/9f9d7705-da3e-45fb-b4fd-3b342766baf8/tuck-in-tales-frontend/src/models/memory.ts
export interface MemorySuggestion {
  type: 'new_character' | 'update_character_bio' | 'update_character_visual' | 'update_character_avatar';
  character_id?: string;
  character_name?: string;
  data: Record<string, any>;
}

export interface Memory {
  id: string;
  family_id: string;
  text: string | null;
  date: string;
  photo_paths: string[];
  categories: string[];
  summary: string | null;
  linked_character_ids: string[];
  analysis_status: 'PENDING' | 'ANALYZING' | 'ANALYZED' | 'CONFIRMED' | 'FAILED';
  raw_analysis: any;
  suggestions: MemorySuggestion[] | null;
  created_at: string;
}

export interface MemoryConfirmRequest {
  categories: string[];
  summary: string | null;
  linked_character_ids: string[];
}

export interface ApplySuggestionRequest {
  suggestion_index: number;
  approved: boolean;
}

export interface MemorySearchRequest {
  query: string;
  match_threshold?: number;
  match_count?: number;
}

export interface MemorySearchResult extends Memory {
  similarity: number;
}

// --- New character detection types ---

export interface NewCharacterDetection {
  name: string;
  guessed_bio: string;
}

export interface PhotoDetectedPerson {
  name: string;
  confidence: 'high' | 'medium' | 'low';
  is_known: boolean;
  visual_note?: string;
  face_x?: number; // 0-100, percentage from left
  face_y?: number; // 0-100, percentage from top
}

export interface PhotoAnalysisResult {
  photo_index: number;
  description: string;
  detected_people: PhotoDetectedPerson[];
}

export interface NewCharacterFromMemoryRequest {
  name: string;
  bio?: string;
  face_x?: number;
  photo_index?: number;
}

export interface CreateCharactersFromMemoryRequest {
  characters: NewCharacterFromMemoryRequest[];
}

export type MemoryCategory = 'context' | 'new_character' | 'important_event' | 'feeling_emotion' | 'watch_for';

export const CATEGORY_LABELS: Record<MemoryCategory, string> = {
  context: 'Context',
  new_character: 'New Character',
  important_event: 'Important Event',
  feeling_emotion: 'Feeling / Emotion',
  watch_for: 'Watch For',
};

export const CATEGORY_COLORS: Record<MemoryCategory, string> = {
  context: 'bg-blue-100 text-blue-800',
  new_character: 'bg-purple-100 text-purple-800',
  important_event: 'bg-amber-100 text-amber-800',
  feeling_emotion: 'bg-pink-100 text-pink-800',
  watch_for: 'bg-red-100 text-red-800',
};
